src/Features/Core/Portable/MakeFieldReadonly/AbstractMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.MakeFieldReadonly
+{
+    internal abstract class AbstractMakeFieldReadonlyDiagnosticAnalyzer<TIdentifierNameSyntax, TConstructorDeclarationSyntax>
+        : AbstractCodeStyleDiagnosticAnalyzer
+        where TIdentifierNameSyntax : SyntaxNode
+        where TConstructorDeclarationSyntax : SyntaxNode
+    {
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(FeaturesResources.Add_readonly_modifier), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(FeaturesResources.Make_field_readonly), WorkspacesResources.ResourceManager, typeof(WorkspacesResources));
+
+        public AbstractMakeFieldReadonlyDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId,
+                   s_localizableTitle, s_localizableMessage)
+        {
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        internal void AnalyzeType(SyntaxNodeAnalysisContext context)
+        {
+            var optionSet = context.Options.GetDocumentOptionSetAsync(context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var option = optionSet.GetOption(CodeStyleOptions.PreferReadonly, context.Node.Language);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            var typeSymbol = (ITypeSymbol)context.SemanticModel.GetDeclaredSymbol(context.Node);
+
+            var nonReadonlyFieldMembers = PooledHashSet<IFieldSymbol>.GetInstance();
+            foreach (var item in typeSymbol.GetMembers())
+            {
+                if (item is IFieldSymbol symbol &&
+                    symbol.DeclaredAccessibility == Accessibility.Private &&
+                    !symbol.IsReadOnly &&
+                    !symbol.IsConst &&
+                    !symbol.IsImplicitlyDeclared &&
```

Comment:
:exclamation: this needs to filter to only include fields in the part declared by `context.Node`. This is the underlying cause of some formatting bugs we observed and I'm already in the process of preparing the fix with tests.

Reply:
➡️ Fixed in 1491c33